python/ql/src/semmle/python/frameworks/Jmespath.qll

```diff
@@ -0,0 +1,35 @@
+/**
+ * Provides classes modeling security-relevant aspects of the `jmespath` PyPI package.
+ * See https://pypi.org/project/jmespath/.
+ */
+
+private import python
+private import semmle.python.dataflow.new.DataFlow
+private import semmle.python.dataflow.new.TaintTracking
+private import semmle.python.Concepts
+private import semmle.python.ApiGraphs
+
+/**
+ * Provides models for the `jmespath` PyPI package.
+ * See https://pypi.org/project/jmespath/.
+ */
+private module Jmespath {
+  class JmespathAdditionalTaintSteps extends TaintTracking::AdditionalTaintStep {
+    override predicate step(DataFlow::Node nodeFrom, DataFlow::Node nodeTo) {
+      exists(DataFlow::CallCfgNode call |
+        call = API::moduleImport("jmespath").getMember("search").getACall() and
+        nodeFrom in [call.getArg(1), call.getArgByName("data")] and
+        nodeTo = call
+        or
+        call =
+          API::moduleImport("jmespath")
+              .getMember("compile")
+              .getReturn()
+              .getMember("search")
+              .getACall() and
+        nodeFrom in [call.getArg(0), call.getArgByName("value")] and
+        nodeTo = call
```

Comment:
You could get rid of the duplication of line 22 by moving `nodeTo = call` into a separate part of the `exists`.

Also, I wonder if you could just change the type of `nodeTo` to `CallCfgNode` and forego the `exists` altogether. (I know overriding predicates can specify a more specific result type, but off the top of my head I don't know if this applies to arguments as well).

Reply:
You're not allowed to provide more specific types for the arguments when overriding.

We could do `exists(DataFlow::CallCfgNode | nodeTo = Call | ...)`, but I feel that the fact that the nodeTo is the _result_ of the call, and not the post-update node of one of the arguments is more accidental than _the way it always must be_ (although it probably will be for +90% of all additional taint steps we write involving calls). I think I'm think I'm just going to leave it as is, but probably wouldn't be too resistant to changing it in the future :wink: